rocketmq/client.py

```diff
@@ -443,25 +455,33 @@ def pull(self, topic, expression='*', max_num=32):
             ctypes.pointer(queue_size)
         ))
         for i in range(int(queue_size.value)):
-            tmp_offset = ctypes.c_longlong()
-            while True:
+            mq = message_queue[i]
+            tmp_offset = ctypes.c_longlong(self.get_message_queue_offset(mq))
+
+            has_new_msg = True
+            while has_new_msg:
                 pull_res = dll.Pull(
                     self._handle,
-                    ctypes.pointer(message_queue[i]),
+                    ctypes.pointer(mq),
                     _to_bytes(expression),
                     tmp_offset,
                     max_num,
                 )
+
                 if pull_res.pullStatus != _CPullStatus.BROKER_TIMEOUT:
                     tmp_offset = pull_res.nextBeginOffset
+                    self.set_message_queue_offset(mq, tmp_offset)
+
                 if pull_res.pullStatus == _CPullStatus.FOUND:
                     for i in range(int(pull_res.size)):
                         yield RecvMessage(pull_res.msgFoundList[i])
-                elif pull_res.pullStatus in [_CPullStatus.NO_MATCHED_MSG, _CPullStatus.NO_NEW_MSG, _CPullStatus.OFFSET_ILLEGAL]:
-                    dll.ReleasePullResult(pull_res)  # NOTE: No need to check ffi return code here
-                    break
+                elif pull_res.pullStatus == _CPullStatus.NO_MATCHED_MSG:
+                    pass
+                elif pull_res.pullStatus == _CPullStatus.NO_NEW_MSG:
+                    has_new_msg = False
+                elif pull_res.pullStatus == _CPullStatus.OFFSET_ILLEGAL:
+                    pass
                 else:
-                    dll.ReleasePullResult(pull_res)  # NOTE: No need to check ffi return code here
-                    break
-                dll.ReleasePullResult(pull_res)  # NOTE: No need to check ffi return code here
+                    pass
+            dll.ReleasePullResult(pull_res)  # NOTE: No need to check ffi return code here
```

Comment:
Indentation is wrong?

Reply:
No, I tested it and it works fine.

Reply:
It will only release last `pull_res` which leads to memory leak?

Reply:
OK, I'll fix this